apps/music/js/metadata/ogg.js

```diff
@@ -39,131 +43,194 @@ var OggMetadata = (function() {
+          LazyLoader.load('js/metadata/vorbis_picture.js').then(() => {
+            return VorbisPictureComment.parsePictureComment(metadata);
+          }).then(resolve);
+        }.bind(this);
+
+        this.readCommentHeader(blobview);
+      }.bind(this));
+    },
 
-    var sum = function(a, b) { return a + b; };
-    var comment_length = Array.reduce(header.segment_table, sum, 0);
+    /**
+     * Read the comment header of an Ogg container.
+     *
+     * @param {BlobView} bloview The audio file being parsed.
+     * @param {Metadata} metadata The (partially filled-in) metadata object.
+     */
+    readCommentHeader: function(blobview) {
+      var metadata = {};
+      var header = this.readPageHeader(blobview);
 
-    return new Promise((resolve, reject) => {
-      page.getMore(page.index, comment_length, (fullpage, err) => {
+      var sum = function(a, b) { return a + b; };
+      this.page_length = Array.reduce(header.segment_table, sum, 0);
+
+      blobview.getMore(blobview.index, this.page_length,
+                       function(fullpage, err) {
         if (err) {
-          reject(err);
-          return;
+          throw new Error(err);
         }
 
-        try {
-          // Look for a comment header from a supported codec
-          var first_byte = fullpage.readByte();
-          var valid = false;
-          switch (first_byte) {
-          case 3:
-            valid = fullpage.readBinaryText(6) === 'vorbis';
-            metadata.tag_format = 'vorbis';
-            break;
-          case 79:
-            valid = fullpage.readBinaryText(7) === 'pusTags';
-            metadata.tag_format = 'opus';
-            break;
+        // Look for a comment header from a supported codec
+        var first_byte = fullpage.readByte();
+        var valid = false;
+        switch (first_byte) {
+        case 3:
+          valid = fullpage.readBinaryText(6) === 'vorbis';
+          metadata.tag_format = 'vorbis';
+          break;
+        case 79:
+          valid = fullpage.readBinaryText(7) === 'pusTags';
+          metadata.tag_format = 'opus';
+          break;
+        }
+        if (!valid) {
+          throw new Error('malformed ogg comment packet');
+        }
+
+        var vendor_string_length = fullpage.readUnsignedInt(true);
+        fullpage.advance(vendor_string_length); // skip libvorbis vendor string
+
+        var num_comments = fullpage.readUnsignedInt(true);
+        var readLoop = function(fullpage, err) {
+          if(err) {
+            throw new Error(err);
           }
-          if (!valid) {
-            reject('malformed ogg comment packet');
-            return;
+          blobview.advance(this.page_length);
+
+          if(!this.readAllComments(fullpage, num_comments, metadata)) {
+            var header = this.readPageHeader(blobview);
+            this.page_length = Array.reduce(header.segment_table, sum, 0);
+
+            blobview.getMore(blobview.index, this.page_length, readLoop);
           }
```

Comment:
good point. But that involve moving this.done() around: it is cleaner it is not the promise resolve.